File: libs/sdk-bindings/bindings-android/lib/src/main/kotlin/breez_sdk_notification/job/ReceivePayment.kt
package breez_sdk_notification.job

import android.content.Context
import breez_sdk.BlockingBreezServices
import breez_sdk.BreezEvent
import breez_sdk.Payment
import breez_sdk_notification.Constants.DEFAULT_PAYMENT_RECEIVED_NOTIFICATION_TEXT
import breez_sdk_notification.Constants.DEFAULT_PAYMENT_RECEIVED_NOTIFICATION_TITLE
import breez_sdk_notification.Constants.NOTIFICATION_CHANNEL_PAYMENT_RECEIVED
import breez_sdk_notification.Constants.PAYMENT_RECEIVED_NOTIFICATION_TEXT
import breez_sdk_notification.Constants.PAYMENT_RECEIVED_NOTIFICATION_TITLE
import breez_sdk_notification.NotificationHelper.Companion.notifyChannel
import breez_sdk_notification.ResourceHelper.Companion.getString
import breez_sdk_notification.SdkForegroundService
import breez_sdk_notification.ServiceLogger
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.Json

@Serializable
data class ReceivePaymentRequest(
    @SerialName("payment_hash") val paymentHash: String,
)

class ReceivePaymentJob(
    private val context: Context,
    private val fgService: SdkForegroundService,
    private val payload: String,
    private val logger: ServiceLogger,
) : Job {
    private var receivedPayment: Payment? = null
    private var breezSDK: BlockingBreezServices? = null

    companion object {
        private const val TAG = "ReceivePaymentJob"
    }

    override fun start(breezSDK: BlockingBreezServices) {
        try {
            this.breezSDK = breezSDK
            val decoder = Json { ignoreUnknownKeys = true }
            val request = decoder.decodeFromString(ReceivePaymentRequest.serializer(), payload)
            val payment = breezSDK.paymentByHash(request.paymentHash)
            if (payment != null) {
                this.receivedPayment = payment
                logger.log(TAG, "Found payment for hash: ${request.paymentHash}", "INFO")
                fgService.onFinished(this)
            }
        } catch (e: Exception) {
            logger.log(TAG, "Failed to call start of receive payment notification: ${e.message}", "WARN")
        }
    }

    override fun onEvent(e: BreezEvent) {
        logger.log(TAG, "Received event $e", "TRACE")
        when (e) {
            is BreezEvent.InvoicePaid -> {
                val pd = e.details
                handleReceivedPayment(pd.paymentHash, pd.amountMsat)
                val payment = try {
                    breezSDK?.paymentByHash(pd.paymentHash)
                } catch (e: Exception) {
                    logger.log(
                        TAG,
                        "Failed to load payment by hash ${pd.paymentHash}: ${e.message}",
                        "WARN"
                    )
                    null
                }
                receivedPayment = payment
            }

            is BreezEvent.Synced -> {
                receivedPayment?.let {
                    logger.log(TAG, "Got synced event for received payment", "INFO")
                    fgService.onFinished(this)
                }
            }

            else -> {}
        }
    }

    override fun onShutdown() {}

    private fun handleReceivedPayment(paymentHash: String, amountMsat: ULong) {
        logger.log(TAG, "Received payment. Payment Hash:${paymentHash}", "INFO")
        val amountSat = amountMsat / 1000u
        notifyChannel(
            context,
            NOTIFICATION_CHANNEL_PAYMENT_RECEIVED,
            getString(
                context,
                PAYMENT_RECEIVED_NOTIFICATION_TITLE,
                DEFAULT_PAYMENT_RECEIVED_NOTIFICATION_TITLE
            ),
            String.format(
                getString(
                    context,
                    PAYMENT_RECEIVED_NOTIFICATION_TEXT,
                    "%d",
                    DEFAULT_PAYMENT_RECEIVED_NOTIFICATION_TEXT
                ), amountSat.toLong()
            )
        )
    }
}

File: libs/sdk-bindings/bindings-swift/Sources/BreezSDK/Task/ReceivePayment.swift
import UserNotifications
import Foundation

struct ReceivePaymentNotificationRequest: Codable {
    let payment_hash: String
}

class ReceivePaymentTask : TaskProtocol {
    fileprivate let TAG = "ReceivePaymentTask"
    
    internal var payload: String
    internal var contentHandler: ((UNNotificationContent) -> Void)?
    internal var bestAttemptContent: UNMutableNotificationContent?
    internal var logger: ServiceLogger
    internal var receivedPayment: Payment? = nil
    internal var breezSDK: BlockingBreezServices? = nil
    
    init(payload: String, logger: ServiceLogger, contentHandler: ((UNNotificationContent) -> Void)? = nil, bestAttemptContent: UNMutableNotificationContent? = nil) {
        self.payload = payload
        self.contentHandler = contentHandler
        self.bestAttemptContent = bestAttemptContent
        self.logger = logger
    }
    
    public func onEvent(e: BreezEvent) {
        switch e {
        case .invoicePaid(details: let details):
            self.logger.log(tag: TAG, line: "Received payment.\nPayment Hash:\(details.paymentHash)", level: "INFO")
            let loadedPayment: Payment?
            do {
                loadedPayment = try self.breezSDK?.paymentByHash(hash: details.paymentHash)
            } catch let e {
                self.logger.log(tag: TAG, line: "Failed to load payment by hash \(details.paymentHash): \(e)", level: "WARN")
                loadedPayment = nil
            }
            receivedPayment = loadedPayment
            break
        case .synced:
            self.logger.log(tag: TAG, line: "got synced event", level: "INFO")
            if self.receivedPayment != nil {
                self.onShutdown()
            }
            break
        default:
            break
        }
    }
    
    func start(breezSDK: BlockingBreezServices) throws {
        do {
            self.breezSDK = breezSDK
            let request = try JSONDecoder().decode(ReceivePaymentNotificationRequest.self, from: self.payload.data(using: .utf8)!)
            let existingPayment = try breezSDK.paymentByHash(hash: request.payment_hash)
            if existingPayment != nil {
                self.receivedPayment = existingPayment
                self.logger.log(tag: TAG, line: "Found payment for hash \(request.payment_hash)", level: "INFO")
                self.onShutdown()
            }
        } catch let e {
            self.logger.log(tag: TAG, line: "Failed to call start of receive payment notification: \(e)", level: "ERROR")
        }
    }
    
    func onShutdown() {
        let successReceivedPayment = ResourceHelper.shared.getString(key: Constants.PAYMENT_RECEIVED_NOTIFICATION_TITLE, validateContains: "%d", fallback: Constants.DEFAULT_PAYMENT_RECEIVED_NOTIFICATION_TITLE)
        let failReceivedPayment = ResourceHelper.shared.getString(key: Constants.PAYMENT_RECEIVED_NOTIFICATION_FAILURE_TITLE, fallback: Constants.DEFAULT_PAYMENT_RECEIVED_NOTIFICATION_FAILURE_TITLE)
        let title = self.receivedPayment != nil ? String(format: successReceivedPayment, self.receivedPayment!.amountMsat/1000) : failReceivedPayment
        self.displayPushNotification(title: title, logger: self.logger, threadIdentifier: Constants.NOTIFICATION_THREAD_PAYMENT_RECEIVED)
    }
}
